### PR TITLE
[ENGG-1925] feat: android emulator support

### PR DESCRIPTION
**Pull request by wrongsahil**
Closes: https://github.com/requestly/requestly/issues/1899

https://github.com/requestly/requestly-desktop-app/assets/16779465/0b3103d9-083a-4712-874a-5d31d83c663e

Related
https://github.com/requestly/requestly/pull/1935

### CODE REVIEW

File: src/renderer/actions/apps/mobile/adb-commands.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+      .then((buffer: any) => buffer.toString("utf8")),
+    ...(options.timeout
+      ? [
+          delay(options.timeout).then(() => {
+            throw new Error(`Timeout for ADB command ${command}`);
+          }),
+        ]
+      : []),
+  ]);
+}
```
Comment (nsrCodes) [inline src/renderer/actions/apps/mobile/adb-commands.ts:59]: rewrite this after adding [`detectUnsettledPromise`](https://github.com/requestly/requestly/blob/9158f99e55517015e4ac2543b9dd7f4b98d40f28/app/src/utils/FunctionUtils.ts#L24-L45) to util.ts

File: src/renderer/actions/apps/mobile/adb-commands.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+export function stringAsStream(input: string) {
+  const contentStream = new stream.Readable();
+  // eslint-disable-next-line no-underscore-dangle
+  contentStream._read = () => {};
+  contentStream.push(input);
+  contentStream.push(null);
+  return contentStream;
+}
```
Comment (nsrCodes) [inline src/renderer/actions/apps/mobile/adb-commands.ts:193]: this should be moved to utils.ts

**Review by nsrCodes — APPROVED** [review]
minor cleanup comments. Let's get this merged 🔥

File: src/renderer/actions/apps/mobile/android.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    console.log("[android-adb:isActive] Device not active");
+    return false;
+  }
+
+  async isActivable() {
+    console.log("adb:isActivable", this.id);
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (wrongsahil) [reply src/renderer/actions/apps/mobile/android.js:54]: ADB devices won't be fetched and shown on the UI if `adb` isn't installed on the system.

File: src/renderer/actions/apps/mobile/android.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+      return {
+        message: needsReboot
+          ? "Restarting device to apply changes. Please wait..."
+          : "",
+      };
+    } catch (err) {
+      console.log(err);
+      this.removeProxy(options.deviceId);
+      await delay(1000);
+      throw err;
```
[the remark this answers is not in the document]
Comment (wrongsahil) [reply src/renderer/actions/apps/mobile/android.js:86]: Yes. What i noticed was that sometimes proxy application and removal takes time. So if we don't delay it, proxy wasn't removed from the emulator

File: src/renderer/actions/apps/mobile/adb-commands.ts [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+//           delay(options.timeout).then(() => {
+//             throw new Error(`Timeout for ADB command ${command}`);
+//           }),
+//         ]
+//       : []),
+//   ]);
+// }
+
+export const ANDROID_TEMP = "/data/local/tmp";
+export const SYSTEM_CA_PATH = "/system/etc/security/cacerts";
```
[the remark this answers is not in the document]
Comment (wrongsahil) [reply src/renderer/actions/apps/mobile/adb-commands.ts]: removed

File: src/renderer/actions/apps/mobile/utils.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+export function delay(durationMs: number): Promise<void> {
+  // eslint-disable-next-line no-promise-executor-return
+  return new Promise((resolve) => setTimeout(resolve, durationMs));
+}
+
+export async function waitUntil<T extends unknown>(
+  delayMs: number,
+  tries: number,
+  test: () => Promise<T>
```
[the remark this answers is not in the document]
Comment (wrongsahil) [reply src/renderer/actions/apps/mobile/utils.ts:20]: Not using anywhere. So removed for now

File: src/renderer/actions/apps/mobile/android.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+          console.log(
+            "[android-adb:activate:setupProxy] Proxy setup output",
+            output.toString()
+          )
+        );
+    } catch (err) {
+      console.error("[android-adb:activate:setupProxy] Error", err);
+      throw new Error("Error setting up proxy");
+    }
+  }
```
[the remark this answers is not in the document]
Comment (wrongsahil) [reply src/renderer/actions/apps/mobile/android.js:134]: Makes sense

File: src/renderer/actions/apps/mobile/adb-commands.ts (hunk 2 above)
Comment (wrongsahil) [reply src/renderer/actions/apps/mobile/adb-commands.ts:87]: Done

**Pull request merged**